### ISSUE TITLE
How to choose a proper popsize？

### ISSUE THREAD

**Issue by nju-luke** [CMA-ES/pycma#140]
Hello, @nikohansen. I noticed that you use `4+3*log(N)`  to calculate the default popsize.
Popsize is matter to the computation. I want to know why such a default value is used.
Thank you very much.

**Comment by nikohansen** [comment]
This population size is close to optimal on (many) unimodal functions. In particular, a smaller population size can _not_ greatly improve the performance per function evaluation (say, be two times faster). On multimodal functions, the situation is much more complicated, see for example [Hansen & Kern 2004](https://scholar.google.fr/scholar?cluster=5920403801438582534) or [Loshchilov et al 2012](https://scholar.google.fr/scholar?cluster=520837764805731680), and we generally recommend something like the IPOP scheme (testing increasing population sizes, see for example [Auger & Hansen 2005](https://scholar.google.fr/scholar?cluster=6757233182925514082)).

**Comment by nju-luke** [comment]
Thanks for the reply. I need some time to understand these things.

**Comment by ShuhuaGao** [comment]
@nikohansen Hi, Niko,
I am new to CMA-ES, but I have played with genetic algorithms before. It seems to me that `4+3*log(N)` is very small population size in a GA. I guess the rationality is  "a small population size λ allows a larger number of generations and therefore usually leads to a faster adaptation of the covariance matrix" stated in [The CMA Evolution Strategy: A Tutorial](https://arxiv.org/abs/1604.00772), which should be reasonable.

I have another question about using **a very large population size** (compared with the default one). The background is that, as you know, GPUs are becoming increasingly popular, and I want to parallelize CMA-ES on a GPU. However, the small population size is entirely a waste of GPU computing power. Nowadays a common NVIDIA GPU has more than 2000 cores, which means we may need tens of thousands of threads to completely feed it (assuming a one-to-one correspondence between a thread and an individual in CMA-ES). Do you have experience with a large population size setting? Or could you recommend some publications in this aspect? I guess some of the recommended parameter values have to be changed when a large population is used.

In other words, I want to trade a smaller generation number with a larger population size, e.g., in a _real-time_ application.  Is CMA-ES suitable for this kind of application? Thank you.

**Comment by nikohansen** [comment]
Publications were given above. The population size is an input parameter one can chose relatively freely. Obviously, on 2000 cores, one can run a population size of 2000, or twice a population size of 1000 in parallel, or all population sizes 8, 16, 32, 64, 128, 250, 500, 1000 in parallel (and restarted), and many other combinations. It is unlikely that the first choice is the best, but that depends on the problem to solve, as can be seen in the above references.

**Comment by nju-luke** [comment]
For the problem of multimode, if the problem converges with a small popsize but does not converge with a big popsize. Is there any reasonable explanation?

**Comment by ShuhuaGao** [comment]
> For the problem of multimode, if the problem converges with a small popsize but does not converge with a big popsize. Is there any reasonable explanation?

@nju-luke It seems all evolutionary algorithms converge slower with an increased population. You could try more generations to see if it indeed converges.

**Comment by ShuhuaGao** [comment]
> Publications were given above. The population size is an input parameter one can chose relatively freely. Obviously, on 2000 cores, one can run a population size of 2000, or twice a population size of 1000 in parallel, or all population sizes 8, 16, 32, 64, 128, 250, 500, 1000 in parallel (and restarted), and many other combinations. It is unlikely that the first choice is the best, but that depends on the problem to solve, as can be seen in the above references.

Thank you, Niko. I will check the above papers.

**Comment by nju-luke** [comment]
> > For the problem of multimode, if the problem converges with a small popsize but does not converge with a big popsize. Is there any reasonable explanation?
> 
> @nju-luke It seems all evolutionary algorithms converge slower with an increased population. You could try more generations to see if it indeed converges.

The problem I have encountered is that the sigma keeps getting bigger with the number of iterations, for example, it exceeds 100. Do you think this will continue to converge?

**Comment by nju-luke** [comment]
> > > For the problem of multimode, if the problem converges with a small popsize but does not converge with a big popsize. Is there any reasonable explanation?
> > 
> > 
> > @nju-luke It seems all evolutionary algorithms converge slower with an increased population. You could try more generations to see if it indeed converges.
> 
> The problem I have encountered is that the sigma keeps getting bigger with the number of iterations, for example, it exceeds 100. Do you think this will continue to converge?

After I set {'tolfacupx': 1e20}, the sigma goes to 1e9 ....

**Comment by nikohansen** [comment]
> The problem I have encountered is that the sigma keeps getting bigger with the number of iterations, for example, it exceeds 100. Do you think this will continue to converge?

That does happen once in a while and can go either way. It would be much easier to figure out when you were to share the resulting figure from `cma.plot()`. Otherwise I would have to have just too many questions to ask.

**Comment by nju-luke** [comment]
> > The problem I have encountered is that the sigma keeps getting bigger with the number of iterations, for example, it exceeds 100. Do you think this will continue to converge?
> 
> That does happen once in a while and can go either way. It would be much easier to figure out when you were to share the resulting figure from `cma.plot()`. Otherwise I would have to have just too many questions to ask.

With the command `cma.plot()`, I got this figure

![cma_plot](https://user-images.githubusercontent.com/17408847/88495568-e9cd5180-cfec-11ea-879b-4989d6b31250.png)

**Comment by nikohansen** [comment]
That's probably the wrong figure, as you wrote

> The problem I have encountered is that the sigma keeps getting bigger with the number of iterations, for example, it exceeds 100.

while in the figure we see three iterations and sigma is ever so slightly decreasing.

**Comment by a-z-e-r-i-l-a** [comment]
is there a rule of thumb for choosing popsize?

**Comment by nikohansen** [comment]
Did you check out the first and second post here?
